### PR TITLE
CSI: Add support for CSIDriver object creation

### PR DESCRIPTION
**Pull request by Madhu-1**
**Description of your changes:**

This will be helpful to list the CSI drivers deployed in the kubernetes cluster. if we have N number of
CSI drivers deployed in the cluster, the user can query `kubectl get csidriver` and get the list of
registered CSI drivers and it requires kube 1.14+ and also the functionality is disabled by default

More info here: ceph/ceph-csi#698

Signed-off-by: Madhu Rajanna madhupr007@gmail.com

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

### CODE REVIEW

**Comment by Madhu-1** [comment]
same as https://github.com/ceph/ceph-csi/pull/698, Added DNM label and `[skip ci]` as this requires cephcsi 1.2.2 release

File: cluster/examples/kubernetes/ceph/common.yaml [hunk 1]
```diff
@@ -1367,6 +1367,9 @@ rules:
   - apiGroups: [""]
     resources: ["nodes"]
     verbs: ["get", "list", "watch"]
+  - apiGroups: ["storage.k8s.io"]
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/common.yaml]: With the change in RBAC I'd suggest we don't backport it to keep the upgrades simple for patch releases. There isn't a need to backport, right?

File: Documentation/ceph-upgrade.md [hunk 2]
```diff
@@ -256,7 +256,7 @@ below, which you should change to match where your images are located.
 ```yaml
     env:
     - name: ROOK_CSI_CEPH_IMAGE
-        value: "quay.io/cephcsi/cephcsi:v1.2.1"
+        value: "quay.io/cephcsi/cephcsi:v1.2.2"
```
Comment (travisn) [inline Documentation/ceph-upgrade.md]: Should the version change to v1.2.2 be in a separate commit or PR from the new settings for CSI driver registration? For example, are there any critical fixes in v1.2.2 so we should backport that?

File: cluster/examples/kubernetes/ceph/operator-openshift.yaml [hunk 3]
```diff
@@ -258,6 +258,9 @@ spec:
         #  value: "9090"
         #- name: CSI_RBD_LIVENESS_METRICS_PORT
         #  value: "9080"
+        # CSIDriver object creation in kubernetes.
+        #- name: CSI_DRIVER_REGISTRATION
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/operator-openshift.yaml]: Why do we need to expose a new setting for this? Why not always register it?

**Comment by Madhu-1** [comment]
@leseb @travisn do I need to keep changes in the upgrade or remove it?

**Comment by Madhu-1** [comment]
I will check why CI failed on kube 1.16

File: pkg/operator/ceph/csi/spec.go [hunk 4]
```diff
@@ -311,5 +315,45 @@ func StartCSIDrivers(namespace string, clientset kubernetes.Interface, ver *vers
 			return fmt.Errorf("failed to create rbd service: %+v\n%+v", err, cephfsService)
 		}
 	}
+
+	if ver.Minor >= provDeploymentSuppVersion {
```
Comment (BlaineEXE) [inline pkg/operator/ceph/csi/spec.go]: Shouldn't this be checking the major version as well?

File: pkg/operator/ceph/csi/spec.go [hunk 5]
```diff
@@ -311,5 +315,45 @@ func StartCSIDrivers(namespace string, clientset kubernetes.Interface, ver *vers
...
+	csiDriver := &k8scsi.CSIDriver{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: k8scsi.CSIDriverSpec{
+			AttachRequired: &attach,
+			PodInfoOnMount: &mountInfo,
+		},
+	}
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```
Comment (BlaineEXE) [inline pkg/operator/ceph/csi/spec.go]: Usage of `RetryOnConflict` doesn't seem to be following the usage instructions from the documentation. 
> RetryOnConflict is used to make an update to a resource when you have to worry about conflicts caused by other code making unrelated updates to the resource at the same time. fn should fetch the resource to be modified, make appropriate changes to it, try to update it, and return (unmodified) the error from the update function. On a successful update, RetryOnConflict will return nil. If the update function returns a "Conflict" error, RetryOnConflict will wait some amount of time as described by backoff, and then try again. On a non-"Conflict" error, or if it retries too many times and gives up, RetryOnConflict will return an error to the caller. 

---

What if the resource needs updated, not just created? What happens if multiple Rook operators with different versions are trying to control the resource.

File: cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml [hunk 6]
```diff
@@ -16,22 +16,6 @@ spec:
     spec:
       serviceAccount: rook-csi-cephfs-provisioner-sa
       containers:
-        - name: csi-attacher
```
Comment (BlaineEXE) [inline cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml]: Should this be left but commented out with a note that it is required for k8s 1.13.x and below?

File: cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml [hunk 7]
```diff
@@ -33,22 +33,6 @@ spec:
           volumeMounts:
             - name: socket-dir
               mountPath: /csi
-        - name: csi-rbdplugin-attacher
```
Comment (BlaineEXE) [inline cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml]: Should this be left but commented out with a note that it is required for k8s 1.13.x and below?

File: cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml (hunk 6 above)
Comment (travisn) [reply cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml]: @Madhu-1 Can you clarify what is needed for the attacher? It's only needed for v1.13? In that case we would still need to create this for 1.13, right?

File: cluster/charts/rook-ceph/values.yaml [hunk 8]
```diff
@@ -81,6 +81,7 @@ csi:
     #image: quay.io/k8scsi/csi-provisioner:v1.3.0
   #snapshotter:
     #image: quay.io/k8scsi/csi-snapshotter:v1.2.0
+  # attacher is required if Kubernetes version is less than 1.14
```
Comment (travisn) [inline cluster/charts/rook-ceph/values.yaml]: instead of saying less than 1.14, we can say only for 1.13, right? CSI isn't supported pre-1.13.

File: cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml (hunk 6 above)
Comment (travisn) [reply cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml]: Is removing the attacher related to registering the csi driver? Seems like it should be a separate PR?

File: pkg/operator/ceph/csi/spec.go (hunk 4 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go]: yes that a good idea, will do it

File: pkg/operator/ceph/csi/spec.go (hunk 5 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go]: This was taken from the cluster-driver-registrar which was earlier doing this job https://github.com/kubernetes-csi/cluster-driver-registrar/blob/master/cmd/csi-cluster-driver-registrar/k8s_register.go#L67. updating is not required as of now.

File: cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml (hunk 6 above)
Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml]: @BlaineEXE  if kube is less than 1.14 we are deploying csi as statefulset https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-sts.yaml, this is not required

@travisn attacher is removed by creating the CSIDriver attacher=false as mentioned here https://kubernetes-csi.github.io/docs/skip-attach.html#skip-attach-with-csi-driver-object

File: cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml (hunk 7 above)
Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml]: same explanation as https://github.com/rook/rook/pull/4172#discussion_r342387683

File: cluster/charts/rook-ceph/values.yaml (hunk 8 above)
Comment (Madhu-1) [reply cluster/charts/rook-ceph/values.yaml]: I will reframe this sentence

File: pkg/operator/ceph/csi/spec.go (hunk 5 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go]: update with for loop instead of `RetryOnConflict`

**Comment by Madhu-1** [comment]
@travisn @BlaineEXE PTAL

File: pkg/operator/ceph/csi/spec.go [hunk 9]
```diff
@@ -311,5 +316,49 @@ func StartCSIDrivers(namespace string, clientset kubernetes.Interface, ver *vers
 			return fmt.Errorf("failed to create rbd service: %+v\n%+v", err, cephfsService)
 		}
 	}
+
+	if ver.Major > KubeMinMajor || ver.Minor >= provDeploymentSuppVersion {
```
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go]: I think you'll need a version check like this:
```
if ver.Major > KubeMinMajor || (ver.Major == KubeMinMajor && ver.Minor >= provDeploymentSuppVersion {
```

File: pkg/operator/ceph/csi/spec.go [hunk 10]
```diff
@@ -196,7 +201,7 @@ func StartCSIDrivers(namespace string, clientset kubernetes.Interface, ver *vers
 	tp.RBDGRPCMetricsPort = getPortFromENV("CSI_RBD_GRPC_METRICS_PORT", DefaultRBDGRPCMerticsPort)
 	tp.RBDLivenessMetricsPort = getPortFromENV("CSI_RBD_LIVENESS_METRICS_PORT", DefaultRBDLivenessMerticsPort)
 
-	if ver.Minor < provDeploymentSuppVersion {
+	if ver.Major > KubeMinMajor || ver.Minor < provDeploymentSuppVersion {
```
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go]: the version check may need an update like mentioned below

File: cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml (hunk 6 above)
Comment (travisn) [reply cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml]: Ok, so we don't need to leave this for 1.13 because we're using the statefulset in that case.

File: pkg/operator/ceph/csi/spec.go (hunk 9 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go]: done

File: pkg/operator/ceph/csi/spec.go (hunk 10 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go]: done

**Comment by Madhu-1** [comment]
@travisn Updated PTAL

File: pkg/operator/ceph/csi/spec.go [hunk 11]
```diff
@@ -311,5 +316,49 @@ func StartCSIDrivers(namespace string, clientset kubernetes.Interface, ver *vers
...
+			Name: name,
+		},
+		Spec: k8scsi.CSIDriverSpec{
+			AttachRequired: &attach,
+			PodInfoOnMount: &mountInfo,
+		},
+	}
+	csidrivers := clientset.StorageV1beta1().CSIDrivers()
+
+	err = wait.PollImmediate(2*time.Second, 15*time.Second, func() (done bool, err error) {
```
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go]: Why do we need to retry this? Most commonly, creating a resource should succeed on the first try.

File: pkg/operator/ceph/csi/spec.go [hunk 12]
```diff
@@ -311,5 +316,49 @@ func StartCSIDrivers(namespace string, clientset kubernetes.Interface, ver *vers
...
+		err = createCSIDriverInfo(clientset, CephFSDriverName)
+		if err != nil {
+			return fmt.Errorf("failed to create CSI driver object for %q: %+v", CephFSDriverName, err)
+		}
+	}
 	return nil
 }
+
+// createCSIDriverInfo Registers CSI driver by creating a CSIDriver object
+func createCSIDriverInfo(clientset kubernetes.Interface, name string) error {
```
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go]: what about adding a check to the integration tests that the csi registration is created?

File: pkg/operator/ceph/csi/spec.go (hunk 11 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go]: removed, it's not needed

File: cluster/examples/kubernetes/ceph/upgrade-from-v1.0-apply.yaml [hunk 13]
```diff
@@ -98,6 +98,12 @@ rules:
   - create
   - update
   - delete
+- apiGroups:
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/upgrade-from-v1.0-apply.yaml]: Patch releases don't typically update the RBAC so i'd be concerned about backporting this. Are you needing to backport this? If not, no need to update this file. If so, we should modify the upgrade guide with steps for this in the patch release section.

Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/upgrade-from-v1.0-apply.yaml]: removed RBAC

File: pkg/operator/ceph/csi/spec.go (hunk 12 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go]: if the CSI driver object is not created we won't be able to attach PVC to an app pod. as we are validating mounting PVC to App pod, this will be covered internally.

**Comment by leseb** [comment]
@Madhu-1 is this ready?

**Comment by Madhu-1** [comment]
@leseb its ready for merging
